AEXS/aex-141.md

```diff
@@ -141,21 +157,19 @@ entrypoint owner(token_id: int) : option(address)
 | owner | option(address) |
 
 ### transfer\(\)
-Transfers NFT with ID `token_id` from the `from` address to the `to` address. Will invoke `NFTReceiver` if `to` address belongs to a contract. If provided, `data` will be submitted with the invocation of `NFTReceiver`. Emits the `Transfer` event.
+Transfers NFT with ID `token_id` from the current owner to the `to` address. Will invoke `IAEX141Receiver.on_aex141_received` if the `to` address belongs to a contract. If provided, `data` will be submitted with the invocation of `IAEX141Receiver.on_aex141_received`. Emits the `Transfer` event.
 
-Should throw if:
-- `Call.caller` is not the current owner, an authorized operator or the approved address for this token;
-- `from` isn't the current owner;
-- `token` isn't a valid token;
-- the invocation of `NFTReceiver` fails.
+Throws if:
+- `Call.caller` is NOT the current owner or NOT approved to transfer on behalf of the owner;
+- `token_id` is NOT a valid token;
+- the invocation of `IAEX141Receiver.on_aex141_received` fails.
```

Comment:
in the Example on Ethereum https://ethereum.org/en/developers/tutorials/how-to-implement-an-erc721-market/ the `openTrade` function puts an item in the escrow of a contract that has previously been approved. This will not be possible using "safe transfer" on aeternity as reentrancy is forbidden, so the `on_aex141_received` can't be called back to the originating contract

Reply:
ok now I get what you mean here. mhh 🤔 

what I currently implemented is "just" a transfer to a contract where sender is tracked and data can be provided. this could definitely be a workaround and would require markets to handle such logic based on the optional data param. see: https://github.com/aeternity/aex141-nft-collection-example/blob/master/contracts/nft-receiver-example/AEX141NFTReceiverExample.aes

how could a possible solution look like to address this or is the way mentioned above fine? feedback from @mradkov would maybe also be good here :)

Reply:
I don't understand how you would solve this?

Reply:
need to verify that with an implementation example. can you maybe summarize the exact use case where you see a problem? I mean the exact order and state of the contract calls in regards to approval and token transfer

Reply:
yes I agree

Reply:
as discussed in DM I still think it's better to avoid unsafe transfers. the typical flow is like you said that a different contract could transfer the NFT to himself and then act on behalf of the user of it which is e.g. typically done in marketplace implementations.

I came up with following example implementation to demonstrate the use of `claim_from_contract` where the nft marketplace contract claims the NFT to itself and stores data about the announced sale:
- https://github.com/aeternity/aex141-nft-simple-marketplace/blob/master/contracts/SimpleNFTMarketplaceAE.aes#L32
- https://github.com/aeternity/aex141-nft-simple-marketplace/blob/master/contracts/CollectionUniqueNFTs.aes#L102

this would be a new entrypoint we would have to propose in the default interface. if we agree on that route I will do so and we can finally proceed here :)

Reply:
solved:
- https://github.com/aeternity/AEXs/pull/148/commits/ebfe98786c5badc358fb310049d74219562d357c

also updated the examples accordingly and provided a basic marketplace example to verify:
- https://github.com/aeternity/aex141-nft-simple-marketplace/blob/master/contracts/SimpleNFTMarketplaceAE.aes#L31